PEP 771: Default Extras for Python Software Packages (#4198)
python/peps: peps/pep-0771.rst

PEP: 771
Title: Default Extras for Python Software Packages
Author: Thomas Robitaille <thomas.robitaille@gmail.com>, Jonathan Dekhtiar <jonathan@dekhtiar.com>

> Sponsor: Pradyun Gedam <pradyunsg@gmail.com>

Author: Hmm this made the linter and build angry, I'll see if I can fix it

> Sponsor: Pradyun Gedam <pradyunsg@gmail.com>

Author: It doesn't seem possible to use placeholders for discussions-to despite PEP 12 - will remove for now

> Sponsor: Pradyun Gedam <pradyunsg@gmail.com>

Reviewer: @astrofrog can you add a link to this: https://github.com/wheel-next/pep_771 for the demo.
I'm almost done for an end-to-end demo of this PEP.

I only have one narrow case left to cover. And it uses all Python standard libs (instead of a custom build backend)

> Sponsor: Pradyun Gedam <pradyunsg@gmail.com>

Author: @DEKHTIARJonathan Thanks! I have already included a link to my demo repo under the reference implementation section - I'll take a look at your demo repo tonight, I have some cases I would like to add. Let's discuss by email and I'll update the demo link once we converge.

Status: Draft
Type: Standards Track
Topic: Packaging

> Created: 24-Sep-2024

Author: Done

> Created: 24-Sep-2024

Author: This was fixed following a comment by @warsaw

* FFT libraries, which also have different implementations (e.g. ``scipy.fft`` and pyFFTW)

With current packaging standards, maintainers have to either
require one of the backends or frontends, or require users
to always specify extras, e.g. ``package[backend]`` and therefore risk users
having an unusable installation if they only install ``package``. Having a
way to specify one or more default backend or frontend and providing a way to
override these defaults would provide a much better experience for users, and
the approach described in this PEP will allow this.

Note that this PEP does not aim to address the issue of disallowing conflicting
or incompatible extras - for example if a package requires exactly one frontend
or backend package. There is currently no mechanism in Python packaging
infrastructure to disallow conflicting or incompatible extras to be installed,
and this PEP does not change that.

Rationale
=========

A number of possible solutions have been extensively and vigorously discussed by
the community for several years, including in `this DPO thread
<https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38>`__

> as well as in numerous issues and pull requests. The solution that will be

Author: I've tried to re-write the Rationale to focus on the positive, and have added a bullet point list for readability.

means that package maintainers can choose whether or not to use it, and it is
flexible enough to accommodate both of the major use cases described in
`Motivation`_.

Specification
=============

``Default-Extra`` Metadata Field
---------------------------------

A new metadata field, ``Default-Extra``, will be added to the `core package
metadata <https://packaging.python.org/en/latest/specifications/core-metadata/#core-metadata>`_.
This field allows package maintainers to define an extra that is
automatically included when a user installs the package without specifying any
extras explicitly::

    Default-Extra: recommended

If multiple default extras are needed, one ``Default-Extra:`` entry
should be provided for each one::

    Default-Extra: backend1
    Default-Extra: backend2
    Default-Extra: backend3

New key in ``[project]`` metadata table
---------------------------------------

A new key will be added to the ``[project]`` table in project metadata as
originally defined in :pep:`621` and now defined in the `PyPA specifications
<https://packaging.python.org/en/latest/specifications/>`_. This key will be named

> ``default-optional-dependencies`` with the following description:

Author: Naming things is hard :sweat_smile: - I've had other people also mention opinions about this, so maybe indeed best to discuss once the PEP is published.

* `TOML <https://toml.io/>`_ type: Array of strings
* Corresponding core metadata field: ``Default-Extra``

> Each string in ``default-optional-dependencies`` should be the name of an extra

Author: I've replaced a number of instances of this (though kept a couple of 'should' where appropriate)

defined in `optional-dependencies
<https://packaging.python.org/en/latest/specifications/pyproject-toml/#dependencies-optional-dependencies>`_,
and each extra in this array will be converted to a matching ``Default-Extra``
entry in the core package metadata. Examples of valid usage which would
produce the example ``Default-Extra`` entries presented in the previous section are:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended"
    ]

and:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "backend1",
        "backend2",
        "backend3"
    ]

Overriding default extras
-------------------------

If extras are explicitly given in a dependency specification, the default
extras are not installed. Otherwise, the default extras are used.

For example, if a package
defines a ``extra1`` default extra as well as a non-default ``extra2``
extra, then if a user was to install the package with::

    pip install package

the ``extra1`` dependency would be included. If the user instead uses::

    pip install package[extra2]

then the ``extra1`` extras would not be installed.

> If the same package is specified multiple times in a dependency tree, then if

Reviewer 2: I find this paragraph difficult to parse and specifying.  Can you rephrase for clarity?

would include only required dependencies and no extras. Note that this requires
no additional specification and is a natural consequence of the rule described
in `Overriding default extras`_.

There are however valid use cases where package maintainers may not want to
provide this. For example, in the case of the multiple possible frontends or
backends, it may be that the package would not be functional without any of the
options. To take a specific example, a package may need either PyQt or PySide to
be installed but will not work if none are provided, so a package maintainer may
therefore not want to provide an option to install the package without any
extras.

Backward Compatibility
======================

All package specification cases valid under :pep:`508` will remain valid.
Therefore, this proposal is fully backward-compatible with existing :pep:`508`
usage.

Once packages start defining default extras, those defaults will only be honored
with recent versions of packaging tools which implement this PEP, but those
packages will remain fully backward-compatible with older packaging tools - with
the only difference that the default extras will not be installed automatically

> for older packaging tools.

Reviewer 2: How about: "... will not be installed automatically when older packaging tools are used." ?

Author: Done!

PEP: 771
Title: Default Extras for Python Software Packages
Author: Thomas Robitaille <thomas.robitaille@gmail.com>, Jonathan Dekhtiar <jonathan@dekhtiar.com>
Sponsor: Pradyun Gedam <pradyunsg@gmail.com>
Status: Draft
Type: Standards Track
Topic: Packaging
Created: 13-Jan-2025

Abstract
========

:pep:`508` specifies a mini-language for
declaring package dependencies. One feature of this language is the ability to
specify *extras*, which are optional components of a distribution that, when
used, install additional dependencies. This PEP proposes a mechanism to allow
one or more extras to be installed by default if none are provided explicitly.

Motivation
==========

Various use cases for default extras and possible solutions in this PEP were discussed

> extensively on `this DPO thread <https://discuss.python.org/t/adding-a-default-extra-require-environment/4898/38>`__.

Author: Thanks!

``default-optional-dependencies`` with the following description:

* `TOML <https://toml.io/>`_ type: Array of strings
* Corresponding core metadata field: ``Default-Extra``

Each string in ``default-optional-dependencies`` must be the name of an extra
defined in `optional-dependencies
<https://packaging.python.org/en/latest/specifications/pyproject-toml/#dependencies-optional-dependencies>`_,
and each extra in this array will be converted to a matching ``Default-Extra``
entry in the core package metadata. Examples of valid usage which would
produce the example ``Default-Extra`` entries presented in the previous section are:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended",
    ]

and:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "backend1",
        "backend2",
        "backend3"
    ]

Overriding default extras
-------------------------

If extras are explicitly given in a dependency specification, the default
extras are not installed. Otherwise, the default extras are used.

For example, if a package
defines an ``extra1`` default extra as well as a non-default ``extra2``

> extra, then if a user was to install the package with::

Reviewer 3: ```suggestion
extra, then if a user were to install the package with::
```

    pip install package

the ``extra1`` dependency would be included. If the user instead uses::

    pip install package[extra2]

> then the ``extra1`` extras would not be installed.

Reviewer 3: ```suggestion
then the ``extra1`` extra would not be installed.
```

If the same package is specified multiple times in an installation command or
dependency tree, the default extras must be installed if any of the instances of
the package are specified without extras. For instance::

    pip install package package[extra2]

should install the default extras.

Note that ``package[]`` would continue to be equivalent to ``package`` and would
not be provided as a way to install without default extras (see the `Rejected
Ideas`_ section for the rationale).

Installing without default extras
---------------------------------

In some cases, package maintainers may want to facilitate installing packages
without any default extras. In this case, as will be shown in more detail in
`How to teach this`_, the best approach is to define an extra which could be
called e.g. ``minimal`` or ``nodefault`` (the naming would be up to the package

> maintainer) which would be an empty set of dependencies. If this extras is

Reviewer 3: ```suggestion
maintainer) which would be an empty set of dependencies. If this extra is
```

must be provided for each one::

    Default-Extra: backend1
    Default-Extra: backend2
    Default-Extra: backend3

New key in ``[project]`` metadata table
---------------------------------------

A new key will be added to the ``[project]`` table in project metadata as
originally defined in :pep:`621` and now defined in the `PyPA specifications
<https://packaging.python.org/en/latest/specifications/>`_. This key will be named
``default-optional-dependencies`` with the following description:

* `TOML <https://toml.io/>`_ type: Array of strings
* Corresponding core metadata field: ``Default-Extra``

Each string in ``default-optional-dependencies`` must be the name of an extra
defined in `optional-dependencies
<https://packaging.python.org/en/latest/specifications/pyproject-toml/#dependencies-optional-dependencies>`_,
and each extra in this array will be converted to a matching ``Default-Extra``
entry in the core package metadata. Examples of valid usage which would
produce the example ``Default-Extra`` entries presented in the previous section are:

.. code-block:: toml

    [project]
    default-optional-dependencies = [

>         "recommended"

Reviewer 4: This is very much a style nit you can ignore, but here and the examples below, shall we encourage the trailing comma? Lots of good reasons in https://www.pythonmorsels.com/trailing-commas-in-python/ such as making diffs smaller when you later add another one, and it makes it easy to re-order etc:

```suggestion
        "recommended",
```

    ]

and:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "backend1",
        "backend2",
        "backend3"
    ]

Overriding default extras
-------------------------

If extras are explicitly given in a dependency specification, the default
extras are not installed. Otherwise, the default extras are used.

For example, if a package

> defines a ``extra1`` default extra as well as a non-default ``extra2``

Reviewer 4: ```suggestion
defines an ``extra1`` default extra as well as a non-default ``extra2``
```

when older packaging tools are used.

The only conceptual backward-compatibility issue to consider is the fact that
this PEP changes extras to no longer be strictly additive, in that specifying
an extra such as ``minimal`` could result in fewer packages being installed.

Security Implications
=====================

There are no known security implications for this PEP.

How to teach this
=================

The rule above regarding only installing default extras when no extras
are explicitly specified, combined with the introduction of the
``Default-Extra:`` keyword and ``default-optional-dependencies`` metadata key
allows us to address several different use cases. Below we take a look at the
two specific use cases raised in the `Motivation`_ section and how package
maintainers should be taught to address these.

Recommended dependencies and minimal installations
--------------------------------------------------

First, we consider the case of packages that want recommended
but not strictly required dependencies installed by default, while also
providing a way to only install the required dependencies.

> In order to do this, a package maintainer would define an extras called
> ``recommended`` containing the recommended but not required dependencies, and
> would choose to have this be included as a default extras:

Reviewer 4: ```suggestion
In order to do this, a package maintainer would define extras called
``recommended`` containing the recommended but not required dependencies, and
would choose to have this be included as default extras:
```

Author: I have fixed this by changing an extras -> an extra

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended"
    ]

    [project.optional-dependencies]
    recommended = [
        "package1",
        "package2"
    ]

In this specific case, a package maintainer may want to allow users to also
install the package without the recommended dependencies, in which case they

> could define an empty extras:

Reviewer 4: ```suggestion
could define empty extras:
```

Author: I have fixed this by changing extras to extra

.. code-block:: toml

    [project.optional-dependencies]
    minimal = []
    recommended = [
        "package1",
        "package2"
    ]

This would then allow users to install ``package[minimal]`` which, since
there would be an extra explicitly specified, would mean the default extras

> does not get installed, and since the ``minimal`` extras is empty, no

Reviewer 4: ```suggestion
do not get installed, and since the ``minimal`` extras is empty, no
```

Author: I have fixed this by changing extras to extra

minimal installations without any extras, but also wants to support having users
specify additional extras without removing the default one. Essentially, they
would want:

* ``package[minimal]`` to give an installation without any extras
* ``package`` to install recommended dependencies (in a ``recommended`` extras)
* ``package[additional]`` to install both recommended and additional dependencies (in an ``additional`` extras)

This could be achieved with e.g:

.. code-block:: toml

    [project]
    default-optional-dependencies = [
        "recommended"
    ]

    [project.optional-dependencies]
    minimal = []
    recommended = [
        "package1",
        "package2"
    ]
    additional = [
        "package[recommended]",
        "package3"
    ]

The ability for a package to reference itself in the extras is supported by
existing Python packaging tools.

Teaching package authors
------------------------

Packages making use of any of the approaches above must ensure that they
properly document the options available to users in terms of installation.

Reference Implementation
========================

> The following branch contains a modified version of the `flit

Reviewer 4: ```suggestion
The following branch contains a modified version of the `Flit
```

<https://flit.pypa.io/en/stable/>`_ package which recognizes the
``default-optional-dependencies`` key in ``pyproject.toml`` and adds the
relevant ``Default-Extra:`` metadata to the build package:

https://github.com/astrofrog/flit/tree/default-extras-pep

In addition, the following branch contains a modified version of `pip
<https://pip.pypa.io/en/stable/>`_ which implements the correct logic for
installing extras when default extras are present:

https://github.com/astrofrog/pip/tree/default-extras-pep

Finally, the following repository makes use of the above two branches to
test out different scenarios and use cases described in this PEP:

https://github.com/astrofrog/pep-771-demo

> The implementations above are proof-of-concepts at this time, do not include any

Reviewer 4: ```suggestion
The implementations above are proofs-of-concept at this time, do not include any
```

tests or documentation in the respective packages, and the existing changes have
not yet been reviewed by the relevant maintainers. Nevertheless, they are
functional enough to allow for interested maintainers to try these out.

Rejected Ideas
==============

> Syntax for unselecting extras

Reviewer 4: Hmm, perhaps "deselect" instead of "unselect"?
```suggestion
Syntax for deselecting extras
```

Deselected is something that was once selected, but is no longer.

Unselected is something that is not selected, and may not have been before.

If so, change the others below as well.

  tree included both ``package[pdf]`` and ``package[-pdf]``, and the rules would
  not be intuitive to users.

* More critically, this would introduce new syntax into dependency specification,
  which means that if any package defined a dependency using the new syntax, it
  and any other package depending on it would no longer be installable by existing
  packaging tools, so this would be a major backward compatibility break.

For these reasons, this alternative was not included in the final proposal.

Adding a special entry in ``extras_require``
--------------------------------------------

A potential solution that has been explored as an alternative to introducing the
new ``Default-Extra`` metadata field would be to make use of an extra with a
'special' name.

One example would be to use an empty string::

    Provides-Extra:
    Requires-Dist: numpy ; extra == ''

The idea would be that dependencies installed as part of the 'empty' extras
would only get installed if another extra was not specified. An implementation
of this was proposed in https://github.com/pypa/setuptools/pull/1503, but it
was found that there would be no way to make this work without breaking

> compatibility with existing usage. For example, packages using setuptools via

Reviewer 4: ```suggestion
compatibility with existing usage. For example, packages using Setuptools via
```

a ``setup.py`` file can do::

    setup(
        ...
        extras_require={'': ['package_a']},
    )

> which is valid and equivalent to having ``package_a`` being defined in
> ``install_requires``, so changing the meaning of the empty string requires would
> break compatibility.

Reviewer 4: ?
```suggestion
which is valid and equivalent to having ``package_a`` being defined in
``install_requires``, so changing the meaning of the empty string would
break compatibility.
```
